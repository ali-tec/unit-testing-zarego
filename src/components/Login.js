import React from 'react';
import "./App.css";
import {useHistory} from "react-router-dom";
import getUser from './GetUsers';

const Login = () => {
    const [name, setName] = React.useState("");
    const [email, setEmail] = React.useState("");
    const [error, setError] = React.useState("");

    const history = useHistory();

    const handleSubmit = (e) => {
        e.preventDefault();
        if(!name || !email) setError("Insert your name & email");
        else {
             getUser()
            .then((data) => {

                const fetchedNames = data.map(user => user.username);
                const fetchedEmails = data.map(user => user.email);

                if(!fetchedNames.includes(name) || !fetchedEmails.includes(email)) {
                    throw ({message: "Not found"});
                }
                else {
                    localStorage.setItem("logged", "yes");
                    history.push("/")
                }
            })

        }
    }

    return (
        <div className="form_container">
            <span className="form_container_header">Login Fom</span>
            <form className="form-login" onSubmit={handleSubmit}>
                <input name="name" type="text" placeholder="Type your name..." className="nameInput" onChange={(e) => setName(e.target.value)} />
                <input name="email" type="email" placeholder="Type your email..." className="emailInput" onChange={(e) => setEmail(e.target.value)} />

                {error && <span style={{color: "red"}}>{error}</span>}
                <button type="submit">Login</button>
            </form>
        </div>

    )
}


export default Login