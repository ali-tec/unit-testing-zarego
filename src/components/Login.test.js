import Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Login from './Login';
import getUser from './GetUsers';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';


const someData = [
    {
        "id": 1,
        "name": "Leanne Graham",
        "username": "Bret",
        "email": "Sincere@april.biz",
        "address": {
          "street": "Kulas Light",
          "suite": "Apt. 556",
          "city": "Gwenborough",
          "zipcode": "92998-3874",
          "geo": {
            "lat": "-37.3159",
            "lng": "81.1496"
          }
        },
        "phone": "1-770-736-8031 x56442",
        "website": "hildegard.org",
        "company": {
          "name": "Romaguera-Crona",
          "catchPhrase": "Multi-layered client-server neural-net",
          "bs": "harness real-time e-markets"
        }
      },
      {
        "id": 2,
        "name": "Ervin Howell",
        "username": "Antonette",
        "email": "Shanna@melissa.tv",
        "address": {
          "street": "Victor Plains",
          "suite": "Suite 879",
          "city": "Wisokyburgh",
          "zipcode": "90566-7771",
          "geo": {
            "lat": "-43.9509",
            "lng": "-34.4618"
          }
        },
        "phone": "010-692-6593 x09125",
        "website": "anastasia.net",
        "company": {
          "name": "Deckow-Crist",
          "catchPhrase": "Proactive didactic contingency",
          "bs": "synergize scalable supply-chains"
        }
      },
]

configure({adapter: new Adapter()});

afterEach(() => {
    // clears the state of every mock object
    jest.clearAllMocks();
  });

  beforeAll(() => {
    jest.mock('axios', () => {
        return {
          __esModule: true,
          default: jest.fn()
        }
      });
  })


//   replaces the localstorage lo que esta en el window (global en node) 
//   con esta mock
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  };
  
  global.localStorage = new LocalStorageMock;

test("button should have text of /Login/", () => {
    // renders the component of login
    const wrapper = shallow(<Login  />);
    // finds all the button tags
    expect(wrapper.find("button").text()).toBe("Login");
})

test("should have two input fields", () => {
    const wrapper = shallow(<Login />);
    // we can also use input find method on classes, ids or anytype of texts 
    // to mark the element
    expect(wrapper.find("input")).toHaveLength(2);
    console.log(wrapper.find("input"))
})

test("should get data from the API", () => {
    const mock = new MockAdapter(axios)
    const data = [...someData];
    mock.onGet('https://jsonplaceholder.typicode.com/users').reply(200, data);
    getUser()
    .then((res) => {
        expect(res).toEqual(someData)
    })
})

test('should set a token in localStorage onSubmit event', () => {
    const fakeEvent = { preventDefault: () => console.log('preventDefault') };
    const wrapper = shallow(<Login />);
    // makes sure that form-login is rendering
    expect(wrapper.find('.form-login').length).toBe(1);
    // simulate 
    wrapper.find('.form-login').simulate('submit', fakeEvent);
    localStorage.setItem("logged", "yes");
    expect(localStorage.getItem("logged")).toBe("yes");
});


test("should match the snapshot", () => {
    const wrapper = shallow(<Login />);
    expect(toJson(wrapper)).toMatchSnapshot();
})
