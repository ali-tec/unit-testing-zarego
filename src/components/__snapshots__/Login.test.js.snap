// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="form_container"
>
  <span
    className="form_container_header"
  >
    Login Fom
  </span>
  <form
    className="form-login"
    onSubmit={[Function]}
  >
    <input
      className="nameInput"
      name="name"
      onChange={[Function]}
      placeholder="Type your name..."
      type="text"
    />
    <input
      className="emailInput"
      name="email"
      onChange={[Function]}
      placeholder="Type your email..."
      type="email"
    />
    <button
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
